# JSON examples and exercise
****
+ get familiar with packages for dealing with JSON
+ study examples with JSON strings and files 
+ work on exercise to be completed and submitted 
****
+ reference: http://pandas.pydata.org/pandas-docs/stable/io.html#io-json-reader
+ data source: http://jsonstudio.com/resources/
****

In [124]:
import pandas as pd

## imports for Python, Pandas

In [125]:
import json
from pandas import json_normalize

## JSON example, with string

+ demonstrates creation of normalized dataframes (tables) from nested json string
+ source: http://pandas.pydata.org/pandas-docs/stable/io.html#normalization

In [126]:
# define json string
data = [{'state': 'Florida', 
         'shortname': 'FL',
         'info': {'governor': 'Rick Scott'},
         'counties': [{'name': 'Dade', 'population': 12345},
                      {'name': 'Broward', 'population': 40000},
                      {'name': 'Palm Beach', 'population': 60000}]},
        {'state': 'Ohio',
         'shortname': 'OH',
         'info': {'governor': 'John Kasich'},
         'counties': [{'name': 'Summit', 'population': 1234},
                      {'name': 'Cuyahoga', 'population': 1337}]}]

In [127]:
# use normalization to create tables from nested element
json_normalize(data, 'counties')

name  population
0        Dade       12345
1     Broward       40000
2  Palm Beach       60000
3      Summit        1234
4    Cuyahoga        1337

In [128]:
# further populate tables created from nested element
json_normalize(data, 'counties', ['state', 'shortname', ['info', 'governor']])

name  population    state shortname info.governor
0        Dade       12345  Florida        FL    Rick Scott
1     Broward       40000  Florida        FL    Rick Scott
2  Palm Beach       60000  Florida        FL    Rick Scott
3      Summit        1234     Ohio        OH   John Kasich
4    Cuyahoga        1337     Ohio        OH   John Kasich

****
## JSON example, with file

+ demonstrates reading in a json file as a string and as a table
+ uses small sample file containing data about projects funded by the World Bank 
+ data source: http://jsonstudio.com/resources/

In [129]:
# load json as string
json.load((open('data/world_bank_projects_less.json')))

[{'_id': {'$oid': '52b213b38594d8a2be17c780'},
  'approvalfy': 1999,
  'board_approval_month': 'November',
  'boardapprovaldate': '2013-11-12T00:00:00Z',
  'borrower': 'FEDERAL DEMOCRATIC REPUBLIC OF ETHIOPIA',
  'closingdate': '2018-07-07T00:00:00Z',
  'country_namecode': 'Federal Democratic Republic of Ethiopia!$!ET',
  'countrycode': 'ET',
  'countryname': 'Federal Democratic Republic of Ethiopia',
  'countryshortname': 'Ethiopia',
  'docty': 'Project Information Document,Indigenous Peoples Plan,Project Information Document',
  'envassesmentcategorycode': 'C',
  'grantamt': 0,
  'ibrdcommamt': 0,
  'id': 'P129828',
  'idacommamt': 130000000,
  'impagency': 'MINISTRY OF EDUCATION',
  'lendinginstr': 'Investment Project Financing',
  'lendinginstrtype': 'IN',
  'lendprojectcost': 550000000,
  'majorsector_percent': [{'Name': 'Education', 'Percent': 46},
   {'Name': 'Education', 'Percent': 26},
   {'Name': 'Public Administration, Law, and Justice', 'Percent': 16},
   {'Name': 'Educatio

In [130]:
# load as Pandas dataframe
sample_json_df = pd.read_json('data/world_bank_projects_less.json')

****
## JSON exercise

Using data in file 'data/world_bank_projects.json' and the techniques demonstrated above,
1. Find the 10 countries with most projects
2. Find the top 10 major project themes (using column 'mjtheme_namecode')
3. In 2. above you will notice that some entries have only the code and the name is missing. Create a dataframe with the missing names filled in.

In [131]:
wb_json_df = pd.read_json('data/world_bank_projects.json')

In [132]:
wb_json_df.groupby('countryshortname')['countryshortname'].count().sort_values(ascending=False).head(10)

countryshortname
China                 19
Indonesia             19
Vietnam               17
India                 16
Yemen, Republic of    13
Nepal                 12
Bangladesh            12
Morocco               12
Mozambique            11
Africa                11
Name: countryshortname, dtype: int64

In [133]:
wb_json = json.load((open('data/world_bank_projects.json')))


## Problem 2.
Load json then create a them_namecode table by normalzing it. 

In [134]:
wb_json_normalized = json_normalize(wb_json, 'mjtheme_namecode', errors='ignore')
wb_json_normalized.groupby('name').count().sort_values(by='code', ascending=False)


code
name                                              
Environment and natural resources management   223
Rural development                              202
Human development                              197
Public sector governance                       184
Social protection and risk management          158
Financial and private sector development       130
                                               122
Social dev/gender/inclusion                    119
Trade and integration                           72
Urban development                               47
Economic management                             33
Rule of law                                     12

Note the empty row with a count of 122, Fix that below by changing
white space to NaN

In [138]:
import numpy as np
stripped = wb_json_normalized.replace(r'^\s*$', np.nan, regex=True)
stripped


code                                      name
0       8                         Human development
1      11                                       NaN
2       1                       Economic management
3       6     Social protection and risk management
4       5                     Trade and integration
...   ...                                       ...
1494   10                         Rural development
1495    9                         Urban development
1496    8                         Human development
1497    5                     Trade and integration
1498    4  Financial and private sector development

[1499 rows x 2 columns]

Now use ffil pad to replace the NaN with the code associated with the empty row.

In [136]:
wb_json_inputed = stripped.fillna(axis=1, method='pad')
wb_json_inputed


code                                      name
0       8                         Human development
1      11                                        11
2       1                       Economic management
3       6     Social protection and risk management
4       5                     Trade and integration
...   ...                                       ...
1494   10                         Rural development
1495    9                         Urban development
1496    8                         Human development
1497    5                     Trade and integration
1498    4  Financial and private sector development

[1499 rows x 2 columns]

## Problem 3.
Repeat the groupby below and see that new counts are generated.

In [137]:
wb_json_inputed.groupby('name').count().sort_values(by='code', ascending=False)


code
name                                              
Environment and natural resources management   223
Rural development                              202
Human development                              197
Public sector governance                       184
Social protection and risk management          158
Financial and private sector development       130
Social dev/gender/inclusion                    119
Trade and integration                           72
Urban development                               47
Economic management                             33
11                                              27
4                                               16
2                                               15
10                                              14
8                                               13
Rule of law                                     12
7                                               11
6                                               10
5                                                5
1                                                5
9                                                3
3                                                3